File: Brainfuck.js
'use strict'

const fs = require('fs')

// Links:
//  * Brainfuck from wiki -> https://ru.wikipedia.org/wiki/Brainfuck#%D0%AF%D0%B7%D1%8B%D0%BA%D0%B8_%D0%BD%D0%B0_%D0%BE%D1%81%D0%BD%D0%BE%D0%B2%D0%B5_Brainfuck
//  * ASCII codes table -> https://snipp.ru/handbk/table-ascii

const BRAINFUCK_ASCII_CODE = {
    Plus: 43, // +
    Minus: 45, // -
    Dot: 46, // .
    Comma: 44, // ,
    Left: 60, // <
    Right: 62, // >
    OBracket: 91, // [
    CBracket: 93, // ]
}

const ALL_BRAINFUCK_COMMANDS = new Set([
    BRAINFUCK_ASCII_CODE.Plus,
    BRAINFUCK_ASCII_CODE.Minus,
    BRAINFUCK_ASCII_CODE.Dot,
    BRAINFUCK_ASCII_CODE.Comma,
    BRAINFUCK_ASCII_CODE.Left,
    BRAINFUCK_ASCII_CODE.Right,
    BRAINFUCK_ASCII_CODE.OBracket,
    BRAINFUCK_ASCII_CODE.CBracket
])

const OUTPUT_DIR = './output/';

function addJavaScriptCode(command, commandCount) {
    switch (command) {
        case BRAINFUCK_ASCII_CODE.Plus:
            return 'plus(' + commandCount + ');\n'
        case BRAINFUCK_ASCII_CODE.Minus:
            return 'minus(' + commandCount + ');\n'
        case BRAINFUCK_ASCII_CODE.Left:
            return 'ptr-= ' + commandCount + ';\n';
        case BRAINFUCK_ASCII_CODE.Right:
            return 'ptr+=' + commandCount + ';\n' + 'arrayExtend(ptr);\n';
        case BRAINFUCK_ASCII_CODE.Dot:
            return 'printCurrentData();\n';
        case BRAINFUCK_ASCII_CODE.Comma:
            return 'data[ptr]=readSymbolFromStdIn();\n'
        case BRAINFUCK_ASCII_CODE.OBracket:
            return 'while( data[ptr] ){ \n'.repeat(commandCount);
        case BRAINFUCK_ASCII_CODE.CBracket:
            return '}\n'.repeat(commandCount);
        default:
            throw new Error(`symbol "${currSymbol}" is ignored`)
    }
}

function main(fileName) {
    console.log(`[Start]: ${fileName}`)

    let javaScriptSrc = "// ------ Java script code for: " + fileName + "\n"
    javaScriptSrc += fs.readFileSync('./brainfuck.translator.template', 'utf8');

    // translate code
    let prevSymbol = undefined
    let prevSymbolCount = 0
    const brainfuckSrc = fs.readFileSync(fileName)
    for (let i = 0; i < brainfuckSrc.length; i++) {
        const currSymbol = brainfuckSrc.readUInt8(i)
        if (!ALL_BRAINFUCK_COMMANDS.has(currSymbol)) {
            // continue insufficient symbol
            console.log(`current symbol "${currSymbol}" is not valid Brainfuck command`)
            continue
        }

        if (prevSymbol === undefined || prevSymbol === currSymbol) {
            // Count repetitions number
            // increase symbol count and move to next symbol
            if (prevSymbol === undefined) {
                prevSymbol = currSymbol
            }
            prevSymbolCount++
            continue
        }

        javaScriptSrc += addJavaScriptCode(prevSymbol, prevSymbolCount)

        // increase counters
        prevSymbol = currSymbol
        prevSymbolCount = 1
    }

    if (prevSymbol !== undefined) {
        javaScriptSrc += addJavaScriptCode(prevSymbol, prevSymbolCount)
    }


    let outputFileName = fileName
        .substring(fileName.lastIndexOf('/') + 1, fileName.length)
        .concat('.js')

    console.log(`java script src code:\n${javaScriptSrc}`)
    if (!fs.existsSync(OUTPUT_DIR)){
        fs.mkdirSync(OUTPUT_DIR);
    }
    fs.writeFileSync(OUTPUT_DIR + outputFileName, javaScriptSrc)

    const f = new Function('require', javaScriptSrc)
    f(require) // require fs to be able to import modules

    console.log(`[End]: ${fileName}`)
}

main('./examples/Hello1.b')
// main('./examples/Hello2.b')
// main('./examples/Beer.b')
// main('./examples/Mandelbrot.b')

File: Brainfuck.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"strings"
)

const BrainfuckPlus = 43     // +
const BrainfuckMinus = 45    // -
const BrainfuckDot = 46      // .
const BrainfuckComma = 44    // ,
const BrainfuckLeft = 60     // <
const BrainfuckRight = 62    // >
const BrainfuckOBracket = 91 // [
const BrainfuckCBracket = 93 // ]

var AllBrainfuckCommands = map[uint8]bool{
	BrainfuckPlus:     true,
	BrainfuckMinus:    true,
	BrainfuckDot:      true,
	BrainfuckComma:    true,
	BrainfuckLeft:     true,
	BrainfuckRight:    true,
	BrainfuckOBracket: true,
	BrainfuckCBracket: true,
}

const JsTemplateFilePath = "./brainfuck.translator.template"
const OutputDir = "./output/"

func readFile(fileName string) string {
	file, err := os.Open(fileName)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err = file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	// Efficient string concatenation taken from
	// https://yourbasic.org/golang/build-append-concatenate-strings-efficiently/
	scanner := bufio.NewScanner(file)
	var strBuilder strings.Builder
	strBuilder.Grow(32)
	for scanner.Scan() { // internally, it advances token based on seperator
		_, err = fmt.Fprintf(&strBuilder, scanner.Text()+"\n") // token in unicode-char
		if err != nil {
			log.Fatal(err)
		}
	}
	return strBuilder.String()
}

func writeFile(fileName string, content string) {
	// create output dif if it's not exist
	if _, err := os.Stat(OutputDir); os.IsNotExist(err) {
		err := os.Mkdir(OutputDir, 0755)
		if err != nil {
			log.Fatal(err)
		}
	}

	// create output file
	outputFileName := fileName[strings.LastIndex(fileName, "/"):] + ".js"
	file, err := os.Create(OutputDir + outputFileName)
	if err != nil {
		log.Fatal(err)
	}

	defer func() {
		if err = file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	// save file to output dir
	_, err2 := file.WriteString(content)
	if err2 != nil {
		log.Fatal(err2)
	}

}

func addJavaScriptCode(command uint8, commandCount int) (string, error) {
	switch command {
	case BrainfuckPlus:
		return fmt.Sprint("plus(", commandCount, ");\n"), nil
	case BrainfuckMinus:
		return fmt.Sprint("minus(", commandCount, ");\n"), nil
	case BrainfuckLeft:
		return fmt.Sprint("if(!ptr) { throw Error('Data pointer can not be negative'); } ptr-= ", commandCount, ";\n"), nil
	case BrainfuckRight:
		return fmt.Sprint("ptr+=", commandCount, ";\n", "arrayExtend(ptr);\n"), nil
	case BrainfuckDot:
		return "printCurrentData();\n", nil
	case BrainfuckComma:
		return "data[ptr]=readSymbolFromStdIn();\n", nil
	case BrainfuckOBracket:
		return strings.Repeat("while( data[ptr] ){ \n", commandCount), nil
	case BrainfuckCBracket:
		return strings.Repeat("}\n", commandCount), nil
	default:
		return "", errors.New(fmt.Sprintf("command '%d' is ignored", command))
	}
}

func translate(fileName string) {
	fmt.Printf("[Start]: %s\n", fileName)

	javaScriptSrc := "// ------ Java script code for: " + fileName + "\n"
	javaScriptSrc += readFile(JsTemplateFilePath)

	// translate code
	brainfuckSrc := strings.TrimSpace(readFile(fileName))
	// brainfuck statement is not empty
	var prevSymbol uint8
	prevSymbolCount := 0
	for i := 0; i < len(brainfuckSrc); i++ {
		currSymbol := brainfuckSrc[i]
		if !AllBrainfuckCommands[currSymbol] {
			// continue insufficient symbol
			fmt.Printf("current symbol %d is not valid Brainfuck command\n", currSymbol)
			continue
		}

		if prevSymbol == 0 || prevSymbol == currSymbol {
			// Count repetitions number
			if prevSymbol == 0 {
				prevSymbol = currSymbol
			}
			prevSymbolCount++
			continue
		}

		code, err := addJavaScriptCode(prevSymbol, prevSymbolCount)
		if err != nil {
			log.Fatal(err)
		}
		javaScriptSrc += code

		// increase counters
		prevSymbol = currSymbol
		prevSymbolCount = 1
	}

	code, err := addJavaScriptCode(prevSymbol, prevSymbolCount)
	if err != nil {
		log.Fatal(err)
	}
	javaScriptSrc += code

	writeFile(fileName, javaScriptSrc)
	fmt.Printf("[End]: %s\n", fileName)
}

func main() {
	translate("./examples/Hello1.b")
	//translate("./examples/Hello2.b")
	//translate("./examples/Beer.b")
	//translate("./examples/Mandelbrot.b")
}
